### PR TITLE
Support parsing `ALTER INDEX TYPE` in Oracle

### PR DESCRIPTION
**Pull request by chakkk309**
Ref https://github.com/apache/shardingsphere/issues/6480 & https://github.com/apache/shardingsphere/issues/18389

- Add  `ALTER INDEX TYPE`  grammar
- Support parsing  `ALTER INDEX TYPE`  in Oracle
- Add test

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+package org.apache.shardingsphere.sql.parser.sql.dialect.statement.oracle.ddl;
+
+import lombok.ToString;
+import org.apache.shardingsphere.sql.parser.sql.common.statement.AbstractSQLStatement;
+import org.apache.shardingsphere.sql.parser.sql.dialect.statement.oracle.OracleStatement;
+
+/**
+ * Oracle alter index type statement.
+ */
+@ToString
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:27]: What is the useful of `@ToString`?

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java (hunk 1 above)
Comment (chakkk309) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:27]: The `@ToString` can override toString() method, but there are no fields in this class. It seems like this annotation is useless here. But I found that almost all SQLStatement class include this `@ToString`, I'm not sure if this has a corresponding design.

Comment (terrymanu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:27]: And there are lots of SQLStatement class exclude `@ToString`. If no useful, please just remove it.

Comment (chakkk309) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:27]: Ok, I will check it.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4 [hunk 2]
```diff
@@ -2735,3 +2735,28 @@ databaseVaultAction
...
+addOrDropClause
+    : (ADD | DROP) operatorName LP_ parameterTypes RP_
+    ;
+
+usingTypeClause
+    : USING implementationType arrayDMLClause?
+    ;
+
+withLocalClause
+    :  (WITH LOCAL RANGE? PARTITION)? storageTableClause?
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/antlr4/imports/oracle/DDLStatement.g4:2752]: Please keep one space after colon.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDDLStatementSQLVisitor.java [hunk 3]
```diff
@@ -901,9 +903,14 @@ public ASTNode visitAlterMaterializedZonemap(final AlterMaterializedZonemapConte
     public ASTNode visitAlterJava(final AlterJavaContext ctx) {
         return new OracleAlterJavaStatement();
     }
-
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDDLStatementSQLVisitor.java:904]: Please execute `mvn spotless:apply` on your local machine to format code.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+package org.apache.shardingsphere.sql.parser.sql.dialect.statement.oracle.ddl;
+
+import org.apache.shardingsphere.sql.parser.sql.common.statement.AbstractSQLStatement;
+import org.apache.shardingsphere.sql.parser.sql.dialect.statement.oracle.OracleStatement;
+
+/**
+ * Oracle alter index type statement.
+ */
+
+public final class OracleAlterIndexTypeStatement extends AbstractSQLStatement implements OracleStatement {
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:27]: @chakkk309 Please add @ToString annotation for OracleAlterIndexTypeStatement. When the user turns on the `sql-show` parameter, the object string should be printed, not the memory address.

Comment (strongduanmu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:27]: In addition, is it possible to add the index name in OracleAlterIndexTypeStatement? In sharding feature, ShardingSphere will rewrite the index name.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDDLStatementSQLVisitor.java (hunk 3 above)
Comment (chakkk309) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDDLStatementSQLVisitor.java:904]: Ok, I will correct the code format.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java (hunk 4 above)
Comment (chakkk309) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:33]: > In addition, is it possible to add the index name in OracleAlterIndexTypeStatement? In sharding feature, ShardingSphere will rewrite the index name.

Do you mean the `index type name`?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDDLStatementSQLVisitor.java (hunk 3 above)
Comment (chakkk309) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/oracle/visitor/statement/impl/OracleDDLStatementSQLVisitor.java]: Ok, I will format the code.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java (hunk 4 above)
Comment (strongduanmu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:33]: @chakkk309 What is the meaning of index type?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Oracle alter index type statement.
+ */
+@Getter
+@Setter
+@ToString
+public final class OracleAlterIndexTypeStatement extends AbstractSQLStatement implements OracleStatement {
+    
+    private IndextypeSegment indextype;
+
```
Comment (strongduanmu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/dialect/statement/oracle/ddl/OracleAlterIndexTypeStatement.java:36]: Please remove useless blank line.

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/ddl/AlterIndexTypeStatementTestCase.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.statement.ddl;
+
+import org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.statement.SQLParserTestCase;
+
+/**
+ * Alter index type statement test case.
+ */
+public final class AlterIndexTypeStatementTestCase extends SQLParserTestCase {
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/ddl/AlterIndexTypeStatementTestCase.java:25]: Please add assert logic for IndexTypeSegment.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/ddl/AlterIndexTypeStatementTestCase.java (hunk 6 above)
Comment (chakkk309) [reply shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/ddl/AlterIndexTypeStatementTestCase.java:34]: It seems like the indextype doesn't have to be parsed, so I'll remove the relevant code.

**Review by strongduanmu — DISMISSED** [review]
LGTM, let's waiting @terrymanu review this pr again.

**Comment by chakkk309** [comment]
There has a Checkstyle violation in `SQLParserTestCases.java` about the FileLength limit :

image

As more and more SQL parsing is supported, the file will become longer and longer. 
And this problem is resolved after pr https://github.com/apache/shardingsphere/pull/19123/ merged.

**Review by strongduanmu — APPROVED** [review]

**Pull request merged**